### ISSUE TITLE
Shape contains "mystery" points?

### ISSUE THREAD

**Issue by mwoehlke-kitware** [valhalla/valhalla#2572]
I'm attempting to route from [38.53566638, -77.32457050](https://www.openstreetmap.org/?mlat=38.53566638&mlon=-77.32457050#map=18/38.53566638/-77.32457050) to [38.53680599, -77.32554177](https://www.openstreetmap.org/?mlat=38.53680599&mlon=-77.32554177#map=18/38.53680599/-77.32554177).

The route I get follows [way 386117098](https://www.openstreetmap.org/way/386117098), as expected. However, the path shape I get back consists of the points:
- 38.53565979, -77.32458496
- 38.53565598, -77.32458496
- 38.53680038, -77.32550812
- 38.53681183, -77.32551575

Where are these points, in particular the first two, coming from? They don't match any geometry that actually exists, and this is messing up my ability to map back to the OSM way. (In particular, it's breaking my logic to figure out what direction the way is being traversed due to an incorrect initial heading being computed.)

The closest nodes are:
- [node 218413049](https://www.openstreetmap.org/node/218413049) (38.5356556, -77.3245873; pretty close to the first two points)
- [node 218413039](https://www.openstreetmap.org/node/218413039) ( 38.5367990, -77.3255080; pretty close to the third point)

I suppose the fourth might be a point along the way "after" the second node, but why do *none* of them match the actual OSM geometry?

**Comment by kevinkreiser** [comment]
Routing in general doesnt only snap to nodes in the graph (intersections of ways) or nodes in the original base data (osm nodes), it will snap to the point along the edge in the graph most close to the input location. Think of the extreme case: a pair of nodes in osm that are separated by 200 meters. If you plop a coordinate in your route request in the middle of that 200 meter stretch, but the router used the closest node that was in osm, then the start of your route would be off of your actual location by about 100 meters. this would be problematic because the route wouldnt start where you are.

You might say ok, what the heck though, why do we have two points at the beginning there? This is because the input location is really close to that osm node. presumably the snapped location is the first coordinate returned and the second coordinate is the coordinate of the first node you linked there. Again we use floating point arithmetic to store this and we use polyline encoding compression as well, so there will be some deviation between the osm node ll and the ll we store in the data.

In summary:

1. the first point is the snapped input coordinate to the geometry of the way (edge)
2. the second point is the first osm node you linked above with some lack of precision from the osm base data

**Comment by mwoehlke-kitware** [comment]
> Routing in general doesnt only snap to nodes in the graph (intersections of ways) or nodes in the original base data (osm nodes), it will snap to the point along the edge in the graph most close to the input location.

Right, I get that. However, a) *none* of the path points match the actual nodes, which is surprising in itself, and b) two of the path points (so, one segment of the path) are almost perfectly due east/west of each other which is at odds with the *actual* direction of the way. The latter in particular is *highly* suspect.

If I just got points that were along the way, in theory I would not be having problems. Getting two points whose relative bearing is so far off from the *actual* bearing of the associated way is what's causing my code to break.

Possibly this is, as noted, due to precision/"compression" artifacts. I still think it's a bug. If I was writing a navigation app, something like this would have my app directing the user to "proceed east" on a road that is running north and south. Maybe Valhalla needs to be clever enough to delete path segments that are so close together that they can't be represented with a reasonable degree of (relative) accuracy?

**Comment by kevinkreiser** [comment]
one of the main points of this project was to build a system that is as portable as possible. this meant we made design decisions that would make tradeoffs between data size, functionality/flexibility, accuracy etc. We decided we would support up to 6 decimal places of accuracy and that we would use floats to store and work on such data. The latter decision is something that we can change in the future the but the former is not something we will change.

What this concretely means is that because osm natively stores 7 decimal places and we store 6 (because of size concerns) is that we will never agree exactly with the source data. The amount of error (sub meter) is deemed acceptable for our use case.

With respect to the two points being in the wrong direction. I'm fairly certain this is a limitation of `float` vs `double`. This is something we could change but its a potentially big task (many changes in test cases is likely the longest part of the work). This is probably something we want to do but we are not committed to doing it at this time. another important point is that if you want to do a bearing computation you should expect some reasonable distance between points to make the result meaningful. measuring the bearing between points who are less than a meter apart is a recipe for inaccuracy.

Lets try to reset here...

You want to know what OSM way and in what direction that way is going? Is that right? Do you want any more information aside from that or is that all you really need? If that is all you need, we have a service (not routing) which does this called `trace_attributes`. Its a map matching algorithm but it gives you back metadata about the underlying map match. Specifically you can have it return the OSM way id and whether or not the edge on the way id was "forward". Forward=true in this case means along the way in the direction the osm nodes were drawn originally (migth be edge cases where oneway=reversed is set on a way).
